spec/codecs/multiline_spec.rb

```diff
@@ -218,4 +214,121 @@
       end
     end
   end
+
+  describe "auto flushing" do
+    let(:config) { {} }
+    let(:codec) { MultilineRspec.new(config).tap {|c| c.register} }
+    let(:events) { [] }
+    let(:lines) do
+      { "en.log" => ["hello world", " second line", " third line"],
+        "fr.log" => ["Salut le Monde", " deuxième ligne", " troisième ligne"],
+        "de.log" => ["Hallo Welt"] }
+    end
+    let(:listener_class) { LineListener }
+    let(:auto_flush_interval) { 0.5 }
+
+    let(:line_producer) do
+      lambda do |path|
+        #create a listener that holds upstream state
+        listener = listener_class.new(events, codec, path)
+        lines[path].each do |data|
+          listener.accept(data)
+        end
+      end
+    end
+
+    context "when auto_flush_interval is not set" do
+      it "does not build any events" do
+        config.update("pattern" => "^\\s", "what" => "previous")
+        line_producer.call("en.log")
+        sleep auto_flush_interval + 0.1
+        expect(events.size).to eq(0)
+        expect(codec.buffer_size).to eq(3)
+      end
+    end
+
+    context "when the auto_flush raises an exception" do
+      let(:errmsg) { "OMG, Daleks!" }
+      let(:listener_class) { LineErrorListener }
+
+      it "does not build any events, logs an error and the buffer data remains" do
+        config.update("pattern" => "^\\s", "what" => "previous",
+          "auto_flush_interval" => auto_flush_interval)
+        codec.logger = MultilineLogTracer.new
+        line_producer.call("en.log")
+        sleep(auto_flush_interval + 0.1)
+        msg, args = codec.logger.trace_for(:error)
+        expect(msg).to eq("Multiline: flush downstream error")
+        expect(args[:exception].message).to eq(errmsg)
+        expect(events.size).to eq(0)
+        expect(codec.buffer_size).to eq(3)
+      end
+    end
+
+    context "mode: previous, when there are pauses between multiline file writes" do
+      it "auto-flushes events from the accumulated lines to the queue" do
+        config.update("pattern" => "^\\s", "what" => "previous",
+          "auto_flush_interval" => auto_flush_interval)
+        line_producer.call("en.log")
+        sleep(auto_flush_interval + 0.1)
+        expect(events[0]).to match_path_and_line("en.log", lines["en.log"])
+        expect(codec).to have_an_empty_buffer
+
+        line_producer.call("fr.log")
+        sleep(auto_flush_interval - 0.3) #next line(s) come before auto-flush i.e. buffered
+        expect(codec.buffer_size).to eq(3)
+        expect(events.size).to eq(1)
+
+        line_producer.call("de.log")
+        sleep(auto_flush_interval + 0.1)
+        expect(events[1]).to match_path_and_line("fr.log", lines["fr.log"])
+        expect(events[2]).to match_path_and_line("de.log", lines["de.log"])
+        expect(codec).to have_an_empty_buffer
```

Comment:
@ph - please tell me which of the two styles you prefer lines 272-286 or lines 310-312